File: 116_populating_next_right_pointers_in_each_node.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    // Queue Version
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    public void connect(TreeLinkNode root) {
        Queue<TreeLinkNode> queue = new LinkedList<>();
        if (root == null) return ;
        
        queue.add(root);
        queue.add(null);
        while ( queue.size() != 0) {
            TreeLinkNode curr = queue.poll();
            if (curr == null) {
                if (queue.size() > 0) queue.offer(null);
            } else {
                curr.next = queue.peek();
                if (curr.left != null) {
                    queue.offer(curr.left);
                }
                if (curr.right != null) {
                    queue.offer(curr.right);
                }
            }
        }
    }
}

File: python/116_populating_next_right_pointers_in_each_node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right, next):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution:
    def connect(sefl, root: 'Node') -> 'Node':
        if not root:
            return None

        curr = root
        next = root.left

        while curr.left:
            curr.left.next = curr.right
            if curr.next:
                curr.right.next = curr.next.left
                curr = curr.next
            else:
                curr = next
                next = curr.left
        return root

    def connect_bfs(self, root: 'Node') -> 'Node':
        q = []
        if root:
            q.append(root)

        while q:
            size = len(q)
            prev = None
            for _ in range(size):
                curr = q.pop(0)
                if not prev:
                    prev = curr
                else:
                    prev.next = curr
                    prev = curr
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
        return root

